src/daemon/input.c

```diff
@@ -150,29 +150,61 @@ static void* play_macro(void* param) {
         }
 
         queued_mutex_unlock(mmutex(kb));
-        // Check if the same combination is still held down.
-        // If the user let go and pressed again, it's okay, because we still need to repeat.
+        // Check if the same combination is still held down, or if it has been released and pressed, again.
         // The other threads will still be queued up and run after this.
         // Maybe in the future we can add an option to prevent new threads with the same combo.
-        // Additionally, wait for a bit, because otherwise short macros will be repeated even if pressed only once
-        if(firstloop){
-            clock_nanosleep(CLOCK_MONOTONIC, 0, &init_repeat_delay, NULL);
-            firstloop = 0;
-        } else {
-            // Give some time for userspace applications to catch up
-            clock_nanosleep(CLOCK_MONOTONIC, 0, &catchup_repeat_delay, NULL);
-        }
         queued_mutex_lock(imutex(kb));
-        const int retrigger = macromask(kb->input.keys, macro->combo);
+        int loopcounter = firstloop ? DELAY_REPEAT_INITIAL : DELAY_REPEAT_CATCHUP;
+        for( ; loopcounter > 0; loopcounter-- ) {
+            const int keydown = macromask(kb->input.keys, macro->combo);
+            // If the user let go and pressed again, macro->triggered will be 1
+            if (macro->triggered == 1) {
+                if (keydown) {
+                    macro->triggered = 2;
+                } else {
+                    macro->triggered = 0;
+                }
+                // Immediately run the macro again, as if first triggered. (see below)
+                firstloop = 2;
+                loopcounter = 0;
+                break;
+            }
+            if (!keydown) {
+                // The key was released.  Stop repeating.
+                // FIXME:  There's a very brief race here:
+                // If the button combination is depressed between now, and the full reset,
+                // the new key press event will be missed.  We accept this very brief window
+                // in the interest of performance and code simplicity.  Mashing the buttons
+                // may miss some events.  Don't use this to count something critical.
+                loopcounter = 1;
+                break;
+            }
+
+            queued_mutex_unlock(imutex(kb));
+            clock_nanosleep(CLOCK_MONOTONIC, 0, &macrodelay, NULL);
+            queued_mutex_lock(imutex(kb));
+        }
         queued_mutex_unlock(imutex(kb));
-        if(!retrigger)
-            break;
+
+        if (!loopcounter) {
+            if (firstloop == 2) {
+                // Use firstloop == 2 as flag to reset firstloop (see above)
+                firstloop = 1;
```

Comment:
Yes, this is another enum.  For `0` and `1` it retains the old meaning (whether or not to have `DELAY_REPEAT_INITIAL` or `DELAY_REPEAT_CATCHUP` between macro re-invocations (if the macro button is consistently held).

The value `2` indicates that the next loop should be treated as if it were the fist loop---because the macro button was released and depressed in the intervening time.

Reply:
I added documentation about this variable (and also simplified it back to the original meaning).